### DISCUSSION TITLE
WebServer enhancer and Socket support in NextJS

### DISCUSSION THREAD

**Discussion by guidiego** [vercel/next.js#18982] [Ideas]
# Feature request
I am studying Socket.IO integration with NEXT.JS and I start to create a PR to contribute with Next in this integration. During my research and code reading, I formulate a proposal and want to discuss before open any PR :) 

## Is your feature request related to a problem? Please describe.
Today I can integrate Next + Socket.io using a custom server. As a NextJS user, I always try to not implement a custom server, I prefer to use all Next power to create my apps/APIs. 

## Describe the solution you'd like
First of all, let's give a quick overview of how SocketIO (I think that is a nice implementation than make all socket logic from zero), works against a web server: 

```typescript
const server = http.createServer(app);
const io = SocketIO({ path: '/ws' }).listen(server);
```

I add the socket to listen to the server, and here we go to our first proposal: **WebServer enhancement option**. Today we only could enhance a web server by creating one from zero. If we take a quick look here:

https://github.com/vercel/next.js/blob/91d4aa54506a68a99f99b8edcfea10a73bea2ec2/packages/next/server/lib/start-server.ts#L13-L19

We only create servers and return the promise resolving incase everything goes well. My idea is to create a ".enhanceWebServer" that gives us the possibility to change some of the server attributes without rewriting the server initialization itself. 

```typescript
  const srv = http.createServer(app.getRequestHandler())
  await new Promise((resolve, reject) => {
    // This code catches EADDRINUSE error if the port is already in use
    srv.on('error', reject)
    srv.on('listening', () => resolve())

    app.enhanceWebServer(srv).then(() => {
       srv.listen(port, hostname)
    })
  })
```

This opens us a window to maybe add on next.config.js a method to enhance and overwrite it without rewriting a custom server! But the core idea here is to use it to lead my second proposal: **Native Socket IO integration on Next.JS** My idea is to create something like:

```typescript
async enhanceWebServer(srv: http.Server) {
     if (this.hasSocketDir) {
       const SocketIO = await import('socket.io');
       const io = SocketIO({ path: '/socket' }).listen(srv);
       const eventsImportPaths = (await fs.readDir(this.socketDir)).filter((file) => file.startsWith('index')))
       const eventsImports = eventsImportPaths.map((file) => import(`${this.socketDir}/${file}`)
       const eventsLogics = Promise.all(eventsImports);
       const events = zip(eventsImportPaths, eventsImports);
       const { connection = noop, disconnect = noop } = await import(this.socketDir);

       io.on('connection', (socket) => {
            connection(io, socket); 
            io.on('diconnect', disconnect);

            Object.keys(events).forEach((eventName) => {
               socket.on(eventName, events[eventName](io, socket);
            });
       });
    }
}
```

## Describe alternatives you've considered
I consider making it a file or working inside the API/pages folder, but I think that does not make sense because the socket will be started with the Server.

## Additional context
How I think in the socket file structure:
```typescript
export default (io: SocketIOClient, socket: SocketIO.Socket) => (param: string) => {
    // event logic
}
```

**Comment by dwisetiyadi** [comment]
This is nice idea. Any progress about this?

**Reply by guidiego** [reply]
I have started on my machine but never go ahead with it. Will try it :D
